contracts/Identity.sol

```diff
@@ -0,0 +1,396 @@
+        0xa0bec69846cdcc8c1ba1eb93be1c5728385a9e26062a73e238b1beda189ac4c9;
+
+    /**
+     * @dev We can store the role to permissions data currently via bitwise
+     * uint256(...[32 bit ANNOUNCER permissions][32 bit OWNER permissions][32 bit NONE permissions])
+     */
+    uint256 private constant ROLE_PERMISSIONS =
+        // Role.OWNER Mask
+        (((1 << uint32(Permission.ANNOUNCE)) |
+            (1 << uint32(Permission.OWNERSHIP_TRANSFER)) |
+            (1 << uint32(Permission.DELEGATE_ADD)) |
+            (1 << uint32(Permission.DELEGATE_REMOVE))) << (uint32(Role.OWNER) * 32)) |
+            // Role.ANNOUNCER Mask
+            ((1 << uint32(Permission.ANNOUNCE)) << (uint32(Role.ANNOUNCER) * 32));
+
+    /*
+     * @dev Modifier that requires that the contract data be initialized
+     */
+    modifier isInitialized() {
+        require(_delegationData().initialized, "Contract not initialized");
+        _;
+    }
+
+    /**
+     * @dev Constructor is only for use if the contract is being used directly
+     *      Construct with address(0x0) for logic contracts
+     * @param owner Address to be set as the owner
+     */
+    constructor(address owner) {
+        _init(owner);
+    }
+
+    /**
+     * @dev Initialize for use as a proxy's logic contract
+     * @param owner Address to be set as the owner
+     */
+    function initialize(address owner) external {
+        // Checks
+        require(_delegationData().initialized == false, "Already initialized");
+        // Effects
+        _init(owner);
+    }
+
+    function _init(address owner) private {
+        _setDelegateRole(owner, Role.OWNER);
+        _delegationData().initialized = true;
+        _delegationData().domainSeparatorHash = keccak256(
+            abi.encode(
+                EIP712_DOMAIN_TYPEHASH,
+                keccak256("Identity"),
+                keccak256("1"),
+                block.chainid,
+                address(this),
+                SALT
+            )
+        );
+    }
+
+    /**
+     * @dev Return the data storage slot
+     *      Slot used to prevent memory collisions for proxy contracts
+     * @return ds delegation storage
+     */
+    function _delegationData() internal pure returns (DelegationStorage storage ds) {
+        bytes32 position = DELEGATION_STORAGE_SLOT;
+        /* solhint-disable no-inline-assembly */
+        assembly {
+            ds.slot := position
+        }
+    }
+
+    /**
+     * @dev Check to see if the role has a particular permission
+     * @param role The Role to test against
+     * @param permission The Permission to test with the role
+     * @return true if the role is assigned the given permission
+     */
+    function doesRoleHavePermission(Role role, Permission permission) public pure returns (bool) {
+        // bitwise (possible) AND (check single permission mask)
+        return ROLE_PERMISSIONS & (((1 << uint32(permission))) << (uint32(role) * 32)) > 0x0;
```

Comment:
x*32 = x<<5. That should save a little gas. Have you tested this uint256 against a short array of unit32s? I wouldn't guess a  << and a * would be cheaper than an array lookup. but they're doing essentially the same thing, so I wouldn't be surprised if it were.

Reply:
Maybe solidity optimizes x*32 to x<<5, in which case this is better.

Reply:
We had tests in our previous repo for gas calculations. Can we add this here (where "here" == this repo)?

Reply:
@shannonwells I tried once to get https://hardhat.org/plugins/hardhat-gas-reporter.html#continuous-integration working quickly and it didn't, but I will add a separate story to make gas checking a part of CI at least (or create a separate suite of tests for it).

@acruikshank An array doesn't work as ROLE_PERMISSIONS is a constant and (currently) constant arrays are not supported.

@acruikshank  I am surprised that x*32 doesn't get optimized and have the same cost, but it does not. Good catch. Fixed in a coming commit.